openff/system/types.py

```diff
@@ -1,6 +1,146 @@
+import json
+from typing import Any
+
+import numpy as np
+from pydantic import BaseModel
+
 from openff.system import unit
 
 
+class _FloatQuantityMeta(type):
+    def __getitem__(self, t):
+        return type("FloatQuantity", (FloatQuantity,), {"__unit__": t})
+
+
+class FloatQuantity(float, metaclass=_FloatQuantityMeta):
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate_type
+
+    @classmethod
+    def validate_type(cls, val):
+        unit_ = getattr(cls, "__unit__", Any)
+        if unit_ is Any:
+            if isinstance(val, float):
+                # input doesn't have a unit, it's just a float-ish type
+                raise ValueError("This needs unit")
+            elif isinstance(val, unit.Quantity):
+                return unit.Quantity(val)
+            else:
+                raise ValueError(
+                    f"Bad input, expected float or pint.Quantity-like, got {type(val)})"
+                )
+        else:
+            unit_ = unit(unit_)
+            if isinstance(val, unit.Quantity):
+                # some custom behavior could go here
+                assert unit_.dimensionality == val.dimensionality
+                # return through converting to some intended default units (taken from the class)
+                return val.to(unit_)
+                # could return here, without converting
+                # (could be inconsistent with data model - heteregenous but compatible units)
+                # return val
+            elif isinstance(val, (float, int)):
+                return val * unit_
+            elif isinstance(val, str):
+                # could do custom deserialization here?
+                return unit.Quantity(val).to(unit_)
+
+
+class QuantityEncoder(json.JSONEncoder):
+    """JSON encoder for unit-wrapped floats and np arrays. Should work
+    for both FloatQuantity and ArrayQuantity"""
+
+    def default(self, obj):
+        if isinstance(obj, unit.Quantity):
+            if isinstance(obj.magnitude, (float, int)):
+                data = obj.magnitude
+            elif isinstance(obj.magnitude, np.ndarray):
+                data = obj.magnitude.tolist()
+            else:
+                # This shouldn't ever be hit if our object models
+                # behave in ways we expect?
+                raise Exception(
+                    f"trying to serialize unsupported type {type(obj.magnitude)}"
+                )
+            return {
+                "val": data,
+                "unit": str(obj.units),
+            }
+        return json.JSONEncoder.default(self, obj)
+
+
+def custom_quantity_encoder(v):
+    return json.dumps(v, cls=QuantityEncoder)
```

Comment:
Might be worth doing something to hide this so as to ensure it's not used outside of its intended scope (which is inside the model, not the entire class) - is this function necessary, or can the returned function be defined inside of the model config?